Add backups article (#350)
prisma/dataguide: content/10-managing-databases/09-backup-considerations.mdx

This system allows you to back up frequently while only recording each change in the system once.  Each backup will only contain changes that have occurred since the last time any backup was run.  This helps keep the size of each incremental backup manageable and allows you to construct different points in time by combining the most recent full backup with various numbers of incremental backups.

One downside of incremental backups is that restoring data can be a bit more complicated.  The longer it's been since a full backup, the more incremental backups you'll have to apply to get to recent changes.  This can take longer that restoring a single full backup and (at most) a single differential backup.

### Write ahead or transaction log backups

> Databases frequently implement safety mechanisms to help recover from system crashes and unsafe shutdowns.  Depending on the system, these may be called write-ahead logs (WAL) or transaction logs.  While these are primarily used for crash recovery purposes, they can be used as a component of a back up strategy to allow for more flexible archiving.

Reviewer: ```suggestion
Databases frequently implement safety mechanisms to help recover from system crashes and unsafe shutdowns.  Depending on the system, these may be called write-ahead logs (WAL) or transaction logs.  While these are primarily used for crash recovery purposes, they can be used as a component of a backup strategy to allow for more flexible archiving.
```

Another reason replication is not a backup is that it doesn't have any obvious ability to restore data.  While you can replicate changes back and forth between servers, replication isn't concerned with keeping previous versions of data and any historic view of your data will likely be lost when the logs are rotated.  This "deficiency" is a reminder that replication is primarily designed to help organizations increase availability and performance rather than as a data preservation tool.

With that being said, replication can be an important component of a backup and disaster recovery plan.  For instance, replication can be useful in the event of a hardware failure on the primary database.  In this case, administrators can quickly promote a replication follower to the leader role and continue to serve client requests to avoid a lengthy restoration procedure.  Some organizations also set up a "delayed" replica which only applies changes after a period of time has elapsed, allowing them to quickly "roll back" the database state by switching to the replica if necessary.

> Another case where replication is often involved in backup processes is as the target of backup operations.  Many times, it's easiest to back up a replica than the primary server.  For instance, to get a consistent file-level backup, you could configure a secondary database as a replica of your production database.  Once the database is synchronized, you can turn off replication temporarily and perform a backup of the replica without affecting your production traffic.  When the backup is complete, you can turn replication back on to resynchronize the server with the changes that occurred during the backup window.

Reviewer: ```suggestion
Another case where replication is often involved in backup processes is as the target of backup operations.  Many times, it's easier to back up a replica than the primary server.  For instance, to get a consistent file-level backup, you could configure a secondary database as a replica of your production database.  Once the database is synchronized, you can turn off replication temporarily and perform a backup of the replica without affecting your production traffic.  When the backup is complete, you can turn replication back on to resynchronize the server with the changes that occurred during the backup window.
```

Reviewer: I think the grammar, in this case, is comparative rather than superlative since it's comparing two things, but I could be wrong, and either is fine.

## How do backups affect security?

Backup strategy intersects with security considerations in a few different ways.

> Backups can help protect against certain security incidents by providing secondary sources of data in scenarios like ransomware attacks.  For instance, if an intruder is able to access and encrypt the contents of your primary database, having access to historical snapshots may give you more options for addressing the situation.  For this to be an option, your backup destination be in a separate security context than your source data to prevent an attacker from impacting your backups as well.

Reviewer: ```suggestion
Backups can help protect against certain security incidents by providing secondary sources of data in scenarios like ransomware attacks.  For instance, if an intruder is able to access and encrypt the contents of your primary database, having access to historical snapshots may give you more options for addressing the situation.  For this to be an option, your backup destination must be in a separate security context than your source data to prevent an attacker from impacting your backups as well.
```

When a new backup occurs, the oldest backup may be deleted or occasionally transferred to longer-term storage.  This gives you access to historical data if you need it, but the archive doesn't remain in your normal backup rotation.  A system like this can help give you options for recovery while minimizing the increase in your total storage over time.

### Schedule and automate the backup process

Once you have decided on a backup rotation, its important to schedule and automate as much of the process as possible.  Ensuring that your backups occur without supervision is an important part of safekeeping your data.

Most backup mechanisms include scheduling components, so it shouldn't require special effort in most cases.  It is important, however, to ensure that you have mechanisms in place to alert you when scheduled backups do not occur.  This might be an email alert when a backup fails or a ping to a chat channel that your organization monitors.

> Automating the backup process may also require you to think about how best to implement your security requirements and access requirements.  Is a pull-based backup system, where the backup target pulls data from your production systems make sense?  What type of access does the backup process need to your systems?  What permissions can you remove to limit the reach and impact of the backup account?  These are the types of questions you'll need to ask yourself while implementing your backup system and especially when automating the process.

Reviewer: ```suggestion
Automating the backup process may also require you to think about how best to implement your security requirements and access requirements.  Does a pull-based backup system, where the backup target pulls data from your production systems, make sense?  What type of access does the backup process need to your systems?  What permissions can you remove to limit the reach and impact of the backup account?  These are the types of questions you'll need to ask yourself while implementing your backup system and especially when automating the process.
```

Reviewer: It seems like this sentence got worded weirdly from what you intended.
